test/integration/api/suite_test.go

```diff
@@ -0,0 +1,648 @@
+	. "github.com/onsi/gomega"
+	"github.com/rancher/wrangler-api/pkg/generated/controllers/apps"
+	"github.com/rancher/wrangler-api/pkg/generated/controllers/core"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/release"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/rancher/harvester/pkg/config"
+	"github.com/rancher/harvester/pkg/server"
+	"github.com/rancher/harvester/pkg/settings"
+	"github.com/rancher/harvester/test/framework/envtest"
+	"github.com/rancher/harvester/test/framework/envtest/printer"
+	"github.com/rancher/harvester/test/framework/fuzz"
+
+	dockerapi "github.com/docker/docker/api/types"
+	dockercontainerapi "github.com/docker/docker/api/types/container"
+	dockerclient "github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/rancher/wrangler/pkg/merr"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	"helm.sh/helm/v3/pkg/strvals"
+	v1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var (
+	testSuiteStartErrChan chan error
+	testCtx               context.Context
+	testCtxCancel         context.CancelFunc
+
+	k8sClientCfg clientcmd.ClientConfig
+	harvester    *server.HarvesterServer
+
+	testMinioContainerID   string
+	testHarvesterNamespace string
+)
+
+func TestAPI(t *testing.T) {
+	defer GinkgoRecover()
+
+	RegisterFailHandler(Fail)
+
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"api suite",
+		[]Reporter{printer.NewlineReporter{}})
+}
+
+var _ = BeforeSuite(func(done Done) {
+	defer close(done)
+
+	testCtx, testCtxCancel = context.WithCancel(context.Background())
+	var err error
+
+	By("starting test environment")
+	k8sClientCfg, err = envtest.Start(GinkgoWriter)
+	if err != nil {
+		GinkgoT().Errorf("failed to start test env, %v", err)
+	}
+
+	By("starting harvester runtime")
+	err = constructHarvesterRuntime()
+	if err != nil {
+		GinkgoT().Errorf("failed to construct harvester runtime, %v", err)
+	}
+
+	By("starting harvester")
+	harvester, err = server.New(
+		testCtx,
+		k8sClientCfg,
+	)
+	if err != nil {
+		GinkgoT().Errorf("failed to create harvester server, %v", err)
+	}
+	testSuiteStartErrChan = make(chan error)
+	go func() {
+		testSuiteStartErrChan <- harvester.Start()
+	}()
+
+	// NB(thxCode): since the start of all controllers is not synchronized,
+	// it cannot guarantee the controllers has been start,
+	// which means the cache(informer) has not ready,
+	// so we give a stupid time sleep to trigger the first list-watch,
+	// and please use the client interface instead of informer interface if you can.
+	select {
+	case <-time.After(15 * time.Second):
+	case err := <-testSuiteStartErrChan:
+		// release the chan
+		close(testSuiteStartErrChan)
+		testSuiteStartErrChan = nil
+		GinkgoT().Errorf("failed to start harvester server, %v", err)
```

Comment:
Are we logging an error when `harvester.Start()` returns nil?
There are some principals for channels:
We usually don't need to close them
A receiver should never close them